helper/semvercon/semvercon_test.go

```diff
@@ -0,0 +1,121 @@
+package semvercon
+
+import (
+	"testing"
+
+	"github.com/hashicorp/go-version"
+)
+
+func TestNewConstraint(t *testing.T) {
+	cases := []struct {
+		input string
+		count int
+		err   bool
+	}{
+		{">= 1.2", 1, false},
+		{"1.0", 1, false},
+		{">= 1.x", 0, true},
+		{">= 1.2, < 1.0", 2, false},
+
+		// Out of bounds
+		{"11387778780781445675529500000000000000000", 0, true},
+
+		// Semver only
+		{">= 1.0beta1", 0, true},
+
+		// No pessimistic operator
+		{"~> 1.0", 0, true},
+	}
+
+	for _, tc := range cases {
+		v, err := NewSemverConstraint(tc.input)
+		if tc.err && err == nil {
+			t.Fatalf("expected error for input: %s", tc.input)
+		} else if !tc.err && err != nil {
+			t.Fatalf("error for input %s: %s", tc.input, err)
+		}
+
+		if len(v) != tc.count {
+			t.Fatalf("input: %s\nexpected len: %d\nactual: %d",
+				tc.input, tc.count, len(v))
+		}
+	}
+}
+
+func TestConstraintCheck(t *testing.T) {
+	cases := []struct {
+		constraint string
+		version    string
+		check      bool
+	}{
+		{">= 1.0, < 1.2", "1.1.5", true},
+		{"< 1.0, < 1.2", "1.1.5", false},
+		{"= 1.0", "1.1.5", false},
+		{"= 1.0", "1.0.0", true},
+		{"1.0", "1.0.0", true},
+
+		// Pre-releases are ordered according to Semver v2
+		{"> 2.0", "2.1.0-beta", true},
+		{"> 2.1.0-a", "2.1.0-beta", true},
+		{"> 2.1.0-a", "2.1.1-beta", true},
+		{"> 2.0.0", "2.1.0-beta", true},
+		{"> 2.1.0-a", "2.1.1", true},
+		{"> 2.1.0-a", "2.1.1-beta", true},
+		{"> 2.1.0-a", "2.1.0", true},
+		{"<= 2.1.0-a", "2.0.0", true},
+		{">= 0.6.1", "1.3.0-beta1", true},
+		{"> 1.0-beta1", "1.0-rc1", true},
+
+		// Meta components are ignored according to Semver v2
+		{">= 0.6.1", "1.3.0-beta1+ent", true},
+		{">= 1.3.0-beta1", "1.3.0-beta1+ent", true},
+		{"> 1.3.0-beta1+cgo", "1.3.0-beta1+ent", false},
+		{"= 1.3.0-beta1+cgo", "1.3.0-beta1+ent", true},
+	}
+
+	for _, tc := range cases {
+		c, err := NewSemverConstraint(tc.constraint)
```

Comment:
Nicer to use:

```
t.Run(tc.constraint, func(t *testing.T) {
  // what you have
})

Reply:
Same else where

Reply:
This is a copy of go-version's constraint_test.go, and I wanted to make my changes easily diffable so the differences in behavior were obvious. I'll make a comment at the top of the file about that so people know.